### PR TITLE
#2976. Add more tests for shorthand equality

### PR DESCRIPTION
**Pull request by sgrekhov**

### CODE REVIEW

File: LanguageFeatures/Static-access-shorthand/equality_A01_t09.dart [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  @override
+  bool operator ==(Object other) {
+    if (other is String) {
+      return other == "hello" || other == "world";
+    } else if (other is int) {
+      return other == 5;
+    }
+    return false;
+  }
+}
```
Comment (eernstg) [inline LanguageFeatures/Static-access-shorthand/equality_A01_t09.dart:34]: Amazing! :grin:

File: LanguageFeatures/Static-access-shorthand/equality_A02_t02.dart [hunk 2]
```diff
@@ -50,18 +52,28 @@ main() {
...
-//                     ^
+  if ({m} != {.intTwo()}) {}
+//            ^
 // [analyzer] unspecified
 // [cfe] unspecified
 
-  if (m != condition ? .values[0] : M.intTwo()) {}
-//                     ^
+  if ({"key": m} != {"key": .values[0]}) {}
+//            ^
```
Comment (eernstg) [inline LanguageFeatures/Static-access-shorthand/equality_A02_t02.dart:66]: I guess it doesn't matter, but for consistency, perhaps?:
```suggestion
//                          ^
```
Or is there a reason why the error would actually be reported at `m`?

File: LanguageFeatures/Static-access-shorthand/equality_A02_t05.dart [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+/// does not change the context type of the second operand, so it would not work
+/// with, for example, `Endian.host == wantBig ? .big : .little`. Here the
+/// second operand is not a `<staticMemberShorthand>`, so it won't have a
+/// shorthand context set, and the context type of the second operand of `==` is
+/// the empty context `_`. (It's neither the static type of the first operand,
+/// nor the parameter type of the first operand's `operator==`.)
+///
+/// @description Checks that it is a compile-time error if a pattern has the
+/// form `== e2` or `!= e2` and `e2` is not a precisely shorthand expression.
+/// Test a class.
```
Comment (eernstg) [inline LanguageFeatures/Static-access-shorthand/equality_A02_t05.dart:18]: But these are not patterns, they are equality expressions. Looks like the description or the test should be adjusted for consistency.

File: LanguageFeatures/Static-access-shorthand/equality_A02_t06.dart [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+/// does not change the context type of the second operand, so it would not work
+/// with, for example, `Endian.host == wantBig ? .big : .little`. Here the
+/// second operand is not a `<staticMemberShorthand>`, so it won't have a
+/// shorthand context set, and the context type of the second operand of `==` is
+/// the empty context `_`. (It's neither the static type of the first operand,
+/// nor the parameter type of the first operand's `operator==`.)
+///
+/// @description Checks that it is a compile-time error if a pattern has the
+/// form `== e2` or `!= e2` and `e2` is not a precisely shorthand expression.
+/// Test a mixin.
```
Comment (eernstg) [inline LanguageFeatures/Static-access-shorthand/equality_A02_t06.dart:18]: Same issue as the previous test.

File: LanguageFeatures/Static-access-shorthand/equality_A02_t07.dart [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+/// does not change the context type of the second operand, so it would not work
+/// with, for example, `Endian.host == wantBig ? .big : .little`. Here the
+/// second operand is not a `<staticMemberShorthand>`, so it won't have a
+/// shorthand context set, and the context type of the second operand of `==` is
+/// the empty context `_`. (It's neither the static type of the first operand,
+/// nor the parameter type of the first operand's `operator==`.)
+///
+/// @description Checks that it is a compile-time error if a pattern has the
+/// form `== e2` or `!= e2` and `e2` is not a precisely shorthand expression.
+/// Test an enum.
```
Comment (eernstg) [inline LanguageFeatures/Static-access-shorthand/equality_A02_t07.dart:18]: Same again.

File: LanguageFeatures/Static-access-shorthand/equality_A02_t08.dart [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+/// ...
+/// This special-casing is only against an immediate static member shorthand. It
+/// does not change the context type of the second operand, so it would not work
+/// with, for example, `Endian.host == wantBig ? .big : .little`. Here the
+/// second operand is not a `<staticMemberShorthand>`, so it won't have a
+/// shorthand context set, and the context type of the second operand of `==` is
+/// the empty context `_`. (It's neither the static type of the first operand,
+/// nor the parameter type of the first operand's `operator==`.)
+///
+/// @description Checks that it is a compile-time error if a pattern has the
```
Comment (eernstg) [inline LanguageFeatures/Static-access-shorthand/equality_A02_t08.dart:16]: And again.

**Review by eernstg — COMMENTED** [review]
Looks good. A couple of tests are testing equality expressions, but the `@description` talks about patterns and it would be good if they can be double checked.

File: LanguageFeatures/Static-access-shorthand/equality_A02_t02.dart (hunk 2 above)
Comment (sgrekhov) [reply LanguageFeatures/Static-access-shorthand/equality_A02_t02.dart]: No. Thank you! Fixed.

File: LanguageFeatures/Static-access-shorthand/equality_A02_t05.dart (hunk 3 above)
Comment (sgrekhov) [reply LanguageFeatures/Static-access-shorthand/equality_A02_t05.dart:18]: Ah, no. The idea was to have patterns. Here and below added `case`. Thank you for noticing!

**Review by sgrekhov — COMMENTED** [review]
Updated. Some expressions became patterns. PTAL.

File: LanguageFeatures/Static-access-shorthand/equality_A02_t08.dart [hunk 7]
```diff
@@ -28,22 +28,22 @@ extension type const ET<T>(T value) {
 }
 
 main() {
-  if ([ET(0)] == const [.new(0)]) {}
+  if ([ET(0)] case == const [ET.new(0)]) {}
```
Comment (eernstg) [inline LanguageFeatures/Static-access-shorthand/equality_A02_t08.dart:31]: Wouldn't this need a similar treatment as the other tests? (That is, the `.new` is not changed to `ET.new`, and the caret is moved to the `.`).

Same thing in line 36, 41, 46.

**Review by eernstg — COMMENTED** [review]
LanguageFeatures/Static-access-shorthand/equality_A02_t08.dart was changed in a different way than the others, was that intended? I couldn't see a reason.

File: LanguageFeatures/Static-access-shorthand/equality_A02_t08.dart (hunk 7 above)
Comment (sgrekhov) [reply LanguageFeatures/Static-access-shorthand/equality_A02_t08.dart]: Added `ET.` to check that there were no bugs and accidentally didn't remove it, sorry. Fixed now.

**Review by eernstg — APPROVED** [review]
LGTM!

**Pull request merged**